lte/cloud/go/services/subscriberdb/servicers/subscriberdb_servicer.go

```diff
@@ -80,20 +75,107 @@ func (s *subscriberdbServicer) ListSubscribers(ctx context.Context, req *lte_pro
 		return nil, err
 	}
 
+	subProtos, nextToken, err := loadSubProtosPage(req.PageSize, req.PageToken, networkID, apnsByName, apnResourcesByAPN)
+	if err != nil {
+		return nil, err
+	}
+
+	digest, noUpdates := s.getDigestAndNoUpdates(req, networkID, apnsByName, apnResourcesByAPN)
+	if noUpdates {
+		subProtos = []*lte_protos.SubscriberData{}
+	}
+
+	listRes := &lte_protos.ListSubscribersResponse{
+		Subscribers:   subProtos,
+		NextPageToken: nextToken,
+		Digest:        digest,
+		NoUpdates:     noUpdates,
+	}
+
+	return listRes, nil
+}
+
+func (s *subscriberdbServicer) getDigestAndNoUpdates(
+	req *lte_protos.ListSubscribersRequest,
+	networkID string,
+	apnsByName map[string]*lte_models.ApnConfiguration,
+	apnResourcesByAPN lte_models.ApnResources,
+) (*lte_protos.SubscribersDigest, bool) {
+
+	digest, noUpdates := req.Digest, false
+	// This functionality is currently placed behind a feature flag.
+	if s.flatDigestEnabled && req.PageToken == "" {
+		digestString, err := getDigest(networkID, apnsByName, apnResourcesByAPN)
+		// If digest generation fails, the error is swallowed to not affect the main functionality.
+		if err != nil {
+			glog.Errorf("Generating digest for subscribers in network of gateway %s failed", networkID)
+		} else {
+			digest = &lte_protos.SubscribersDigest{Md5Base64Digest: digestString}
+			noUpdates = req.Digest.Md5Base64Digest == digestString
+		}
+	}
+
+	return digest, noUpdates
+}
+
+// getDigest loads all subscribers registered on the current network, and returns
+// a deterministic subscriber flat digest.
+func getDigest(
+	networkID string,
+	apnsByName map[string]*lte_models.ApnConfiguration,
+	apnResourcesByAPN lte_models.ApnResources,
+) (string, error) {
+	subProtosById := map[string]proto.Message{}
+	curToken := ""
+	for {
```

Comment:
Nit: generally speaking I prefer not using while loops, as the exit condition is implicit / requires reading the whole block to understand

That said, I get that this is basically a "do-for" loop, so this isn't a bad way of handling it. Maybe try a `for curToken != "" || isFirstLoop)`

Reply:
Will be using the `curToken = ""; nextToken = "start"; for nextToken != ""` strategy for this one @hcgatewood